### ISSUE TITLE
extended_trace=true fails for NM

### ISSUE THREAD

**Issue by RossBoylan** [JuliaOpt/Optim.jl#203]
```
julia> @time sol2 = optimize(RB.mylike, RB.troubleParamsInit.raw, iterations=5, store_trace=true, extended_trace=true)  # Nelder Mead
 Iter     Function value   Gradient norm
 ERROR: UndefVarError: x not defined
  [inlined code] from /home/ross/.julia/v0.4/Optim/src/nelder_mead.jl:28
  in optimize at /home/ross/.julia/v0.4/Optim/src/nelder_mead.jl:89
  in optimize at /home/ross/.julia/v0.4/Optim/src/optimize.jl:25
```
The relevant code is in the macro definition (in nelder_mead.jl)
```
macro nmtrace()
    quote
        if tracing
            dt = Dict()
            if o.extended_trace
                dt["x"] = copy(x)
            end
            grnorm = NaN
            update!(tr,
                    iteration,
                    f_x,
                    grnorm,
                    dt,
                    o.store_trace,
                    o.show_trace,
                    o.show_every,
                    o.callback)
        end
    end
end

```
It's not immediately obvious to me what should go in place of x in copy(x).

**Comment by pkofod** [comment]
Can't find the issue, but this is unfortunately known. Thanks for reporting the bug though.

The reason why this isn't fixed, I guess, is because it is related to the fact that we have not decided what to do with regards to the "function value" and "gradient norm" columns. What value do you report for the function value, and could we print something "like" a gradient for NM? I guess we havn't decided what to do yet, and maybe we (want to help? :)) should take a look at the literature and other software.

**Comment by RossBoylan** [comment]
I'm not sure a complete trace would fit easily in a table (which seems to be the current model for recording the trace), since there are individual function evaluations at individual points, calculation of the NM objective function, and various intermediate guesses.  I think ideally both the NM objective function, nmobjective, and the user-supplied object function, values would be recorded.

For the gradient, my initial reaction was to skip it since it's not relevant or otherwise calculated.  Second, also quick, thought would be to use (y_h - y_l)/norm(h-l).

Possibly recording the centroid of the active region would be of interest.

I am not an optimization expert.
Ross
________________________________

**Comment by pkofod** [comment]
I found the discussion: https://github.com/JuliaOpt/Optim.jl/pull/103 (thought it was an issue not a pull).

I know the gradient is not relevant, it's just that the tracing code is not aware of which algorithms is currently being used, so there's not really that much flexibility currently. That's also what you see when the tracing algorithms thinks there's an "x" value to store (although the algorithm's state isn't just  a point called x). The centroid is a possibility, but please see the other discussion to see what has been discussed.

**Comment by RossBoylan** [comment]
I'd expect a full trace to include things like whether a point was generated by reflection, expansion or contraction.  So it seems to me trace info will differ with the type of optimization routine.  The current approach does have a dictionary into which anything in principle could be stored, but it seems more natural  to have a different types for the trace as appropriate.  Then each can have an appropriate method to show it.

#103 and the discussion here has been mostly along the lines of "we're reporting f, x, and the gradient, what values should go in each slot."  It would be better to record the relevant information for the optimizer in use, and then debate what the prinout should display :)

**Comment by pkofod** [comment]
I agree, and I'd be happy to take a shot at it.

**Comment by pkofod** [comment]
closed by #204
